Classes/common/touchdb/TD_Database.m

```diff
@@ -193,47 +229,150 @@ - (BOOL)initialize:(NSString*)updates inDatabase:(FMDatabase*)db
 }
 
 // callers: -open, -compact
-- (BOOL)openFMDB
+- (BOOL)openFMDBWithEncryptionKeyProvider:(id<CDTEncryptionKeyProvider>)provider
 {
-    int flags = SQLITE_OPEN_FILEPROTECTION_COMPLETEUNLESSOPEN;
-    if (_readOnly)
-        flags |= SQLITE_OPEN_READONLY;
-    else
-        flags |= SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE;
-    CDTLogInfo(CDTDATASTORE_LOG_CONTEXT, @"Open %@ (flags=%X)", _path, flags);
-    _fmdbQueue = [FMDatabaseQueue databaseQueueWithPath:_path flags:flags];
-    if (!_fmdbQueue) return NO;
+    __block BOOL result = YES;
+    
+    // Cipher/Not cipher the database
+    NSString* key = [provider encryptionKey];
+    if (!key) {
+        CDTLogDebug(CDTDATASTORE_LOG_CONTEXT, @"No key provided. Is DB at path %@ encrypted?",
+                    _path);
+
+        if ([TD_Database isDatabaseUnencryptedAtPath:_path]) {
+            CDTLogDebug(CDTDATASTORE_LOG_CONTEXT,
+                        @"DB at path %@ is not encrypted or it does not exist. No key is necessary",
+                        _path);
+        } else {
+            CDTLogError(CDTDATASTORE_LOG_CONTEXT,
+                        @"No key provided, however the database is encrypted");
+
+            result = NO;
+        }
+    }
+    
+    // Create database
+    FMDatabaseQueue* queue = nil;
+    
+    if (result) {
+#ifdef SQLITE_OPEN_FILEPROTECTION_COMPLETEUNLESSOPEN
+        int flags = SQLITE_OPEN_FILEPROTECTION_COMPLETEUNLESSOPEN;
+#else
+        int flags = kNilOptions;
+#endif
+        if (_readOnly) {
+            flags |= SQLITE_OPEN_READONLY;
+        } else {
+            flags |= SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE;
+        }
+        CDTLogDebug(CDTDATASTORE_LOG_CONTEXT, @"Open %@ (flags=%X)", _path, flags);
+        
+        queue = [FMDatabaseQueue databaseQueueWithPath:_path flags:flags];
+        result = (queue != nil);
+    }
+
+    // Cipher database
```

Comment:
Comment seems inaccurate -- this checks we can use the key provided with the database file?